File: Core/src/main/java/jokatu/game/viewresolver/ViewResolverFactory.java
package jokatu.game.viewresolver;

import jokatu.game.Game;
import jokatu.game.player.Player;
import org.jetbrains.annotations.NotNull;

import java.text.MessageFormat;

/**
 * A factory for {@link ViewResolver}s.
 * @author steven
 */
public abstract class ViewResolverFactory<P extends Player, G extends Game<P>> {

	@NotNull
	public final ViewResolver<P, G> getViewResolver(@NotNull Game<?> game) {
		G castGame = castGame(game);
		return getResolverFor(castGame);
	}

	@NotNull
	private G castGame(@NotNull Game<?> game) {
		Class<G> gameClass = getGameClass();
		if (!gameClass.isInstance(game)) {
			throw new IllegalArgumentException(
					MessageFormat.format("Game was not of the right type.  Expected {0}.", gameClass)
			);
		}
		return gameClass.cast(game);
	}

	@NotNull
	protected abstract Class<G> getGameClass();

	@NotNull
	protected abstract ViewResolver<P, G> getResolverFor(@NotNull G castGame);
}

File: Core/src/main/kotlin/jokatu/game/viewresolver/ViewResolverFactory.kt
package jokatu.game.viewresolver

import jokatu.game.Game
import jokatu.game.player.Player
import kotlin.reflect.KClass
import kotlin.reflect.full.cast

/**
 * A factory for [ViewResolver]s.
 * @author steven
 */
abstract class ViewResolverFactory<P : Player, G : Game<P>>(protected val gameClass: KClass<G>) {
	fun getViewResolver(game: Game<*>): ViewResolver<P, G> {
		val castGame = castGame(game)
		return getResolverFor(castGame)
	}

	private fun castGame(game: Game<*>): G {
		val gameClass = gameClass
		if (!gameClass.isInstance(game)) {
			throw IllegalArgumentException(
					"Game was not of the right type.  Expected ${gameClass.simpleName} but was ${game.javaClass.simpleName}."
			)
		}
		return gameClass.cast(game)
	}

	protected abstract fun getResolverFor(castGame: G): ViewResolver<P, G>
}
